File: lb.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>

#include "hdf5.h"

#define _USE_MATH_DEFINES
#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#define OUTFILE_NAME "out.h5"
#define DSET_NAME "velocity"
#define RANK 3

static const int cx = N / 4;
static const int cy = M / 2;
static const int r = M / 9;

static const double w[9] = {
   4./9. ,
   1./9. ,
   1./9. ,
   1./9. ,
   1./36.,
   1./36.,
   1./9. ,
   1./36.,
   1./36.
};

static const int c[9][2] = {
   { 0,  0},
   { 0, -1},
   { 0,  1},
   {-1,  0},
   {-1, -1},
   {-1,  1},
   { 1,  0},
   { 1, -1},
   { 1,  1}
};

static const int noslip[9] = {0, 2, 1, 6, 8, 7, 3, 5, 4};

static inline int idx(int i, int j) {
   int id = M * i + j;
   return id;
}

static inline int mod(int a, int b) {
   return (b + a%b) % b;
}

static inline double inlet_vel(int i) {
   double x = (double)i / (M-1);
   return ULB * (1. + 1.e-4 * sin(2.*M_PI * x));
}

static inline double rho(int idx, double* restrict f) {
   double sum = 0;
   for (int q = 0; q < 9; q++)
      sum += f[9*idx + q];

   return sum;
}

static inline double u(int k, int idx, double* restrict f) {
   double sum = 0;
   for (int q = 0; q < 9; q++)
      sum += f[9*idx + q] * c[q][k];

   return sum / rho(idx, f);
}

void boundary(double* restrict f) {
   // outflow
   for (int j = 0; j < M; j++)
      for (int q = 3; q < 6; q++)
         f[9*idx(N-1, j) + q] = f[9*idx(N-2, j) + q];

   // inflow
   for (int j = 0; j < M; j++) {
      double ux = inlet_vel(j);
      double u2 = ux*ux;
      double sum1 = 0, sum2 = 0;
      for (int q = 0; q < 6; q++) {
         if (q < 3)
            sum2 += f[9*j + q];
         else
            sum1 += f[9*j + q];
      }

      double rho_j = 1. / (1 - ux) * (sum2 + 2*sum1);

      for (int q = 6; q < 9; q++) {
         double cu = c[q][0] * ux;
         f[9*j + q] = w[q] * rho_j * (1 + 3 * cu + 0.5 * 9 * cu*cu - 0.5 * 3 * u2);
      }
   }
}

void collstream(double* restrict fnew, double* restrict fold, int* restrict obstacle, double omega) {
#pragma omp parallel for collapse(2) schedule(static)
   for (int i = 0; i < N; i++) {
      for (int j = 0; j < M; j++) {
         int id = idx(i,j);
         if (!obstacle[id]) {
            double ux = 0;
            double uy = 0;
            double rho_ij = 0;
            for (int q = 0; q < 9; q++) {
               rho_ij += fold[9*id + q];
               ux += fold[9*id + q] * c[q][0];
               uy += fold[9*id + q] * c[q][1];
            }
            ux /= rho_ij;
            uy /= rho_ij;
            double u2 = ux*ux + uy*uy;
            for (int q = 0; q < 9; q++) {
               int idnew = 9*idx(mod(i + c[q][0], N), mod(j + c[q][1], M)) + q;
               double cu = c[q][0] * ux + c[q][1] * uy;
               double feq_ijq = w[q] * rho_ij * (1 + 3 * cu + 0.5 * 9 * cu*cu - 0.5 * 3 * u2);
               fnew[idnew] = (1 - omega) * fold[9*id + q] + omega*feq_ijq;
            }
         } else {
            for (int q = 0; q < 9; q++) {
               int idnew = 9*idx(mod(i + c[q][0], N), mod(j + c[q][1], M)) + q;
               fnew[idnew] = fold[9*id + noslip[q]];
            }
         }
      }
   }
}

void update(double* restrict fnew, double* restrict fold, int* restrict obstacle, double omega) {
   collstream(fnew, fold, obstacle, omega);
   boundary(fnew);
}

static herr_t write_u(hid_t dataset, double* restrict f) {
   static double vel[N*M];

   static hsize_t dims[] = {N,M}, start[RANK] = {0}, count[] = {1,N,M};
   hid_t mspace, fspace;
   herr_t err;

   mspace = H5Screate_simple(RANK-1, dims, NULL);
   fspace = H5Dget_space(dataset);

#pragma omp parallel for schedule(static)
   for (int idx = 0; idx < N*M; idx++) {
      double ux = u(0, idx, f);
      double uy = u(1, idx, f);
      vel[idx] = sqrt(ux*ux + uy*uy);
   }

   err = H5Sselect_hyperslab(fspace, H5S_SELECT_SET, start, NULL, count, NULL);
   err = H5Dwrite(dataset, H5T_NATIVE_DOUBLE, mspace, fspace, H5P_DEFAULT, vel);

   start[0]++;

   return err;
}

int main() {
   hid_t file, dataset, dataspace;
   hsize_t dims[] = {T/INTERVAL, N, M};

   file = H5Fcreate(OUTFILE_NAME, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
   dataspace = H5Screate_simple(RANK, dims, NULL);
   dataset = H5Dcreate2(file, DSET_NAME, H5T_NATIVE_DOUBLE, dataspace, H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);

   double nu = ULB * r / RE;
   double omega = 1. / (3. * nu + 0.5);

   double* restrict fnew = malloc(9 * N * M * sizeof(double));
   double* restrict fold = malloc(9 * N * M * sizeof(double));

   int* restrict obstacle = malloc(N*M * sizeof(int));
   for (int i = 0; i < N; i++)
      for (int j = 0; j < M; j++)
         obstacle[idx(i,j)] = ((i-cx)*(i-cx) + (j-cy)*(j-cy) < r*r);

   for (int i = 0; i < N; i++) {
      for (int j = 0; j < M; j++) {
         double ux = inlet_vel(j);
         double u2 = ux*ux;
         for (int q = 0; q < 9; q++) {
            double cu = c[q][0] * ux;
            fold[9*idx(i,j) + q] = w[q] * 1.0 * (1 + 3 * cu + 0.5 * 9 * cu*cu - 0.5 * 3 * u2);
         }
      }
   }

   for (int t = 0; t < T; t++) {
      if (t % 2) {
         update(fold, fnew, obstacle, omega);
      } else {
         if (!(t % INTERVAL)) {
            write_u(dataset, fold);
         }
         update(fnew, fold, obstacle, omega);
      }
   }

   H5Sclose(dataspace);
   H5Dclose(dataset);
   H5Fclose(file);

   free(fnew);
   free(fold);
   free(obstacle);

   return 0;
}

File: lb.f90
module constants
real, parameter :: cx = N / 4.0
real, parameter :: cy = M / 2.0
real, parameter :: r = M / 9.0

real(kind=8), dimension(9) :: w = (/&
   4.0d0/9.0d0 ,&
   1.0d0/9.0d0 ,&
   1.0d0/9.0d0 ,&
   1.0d0/9.0d0 ,&
   1.0d0/36.0d0,&
   1.0d0/36.0d0,&
   1.0d0/9.0d0 ,&
   1.0d0/36.0d0,&
   1.0d0/36.0d0 &
   /)

integer, dimension(9,2) :: c = transpose(reshape((/&
   0,  0,&
   0, -1,&
   0,  1,&
  -1,  0,&
  -1, -1,&
  -1,  1,&
   1,  0,&
   1, -1,&
   1,  1 &
   /), (/2, 9/)))

integer, dimension(9) :: noslip = (/1, 3, 2, 7, 9, 8, 4, 6, 5/)

end module

function pmod(a, b) result(res)
   implicit none
   integer, intent(in) :: a, b
   integer :: res

   res = mod(b + mod(a,b), b)

end function

function inlet_vel(i) result(v)
   implicit none
   integer :: i
   real(kind=8) :: x, v

   x = real(i,8) / (M-1)
   v = ULB * (1 + 0.0001d0 * sin(2 * 3.142d0 * x))

end function

function rho(i, j, f) result(s)
   implicit none
   real(kind=8), dimension(9, N, M) :: f
   integer :: i, j, q
   real(kind=8) :: s

   s = 0.0d0
   do q=1,9
      s = s + f(q, i, j)
   end do

end function

function u(k, i, j, f) result(s)
   use constants, only : c
   implicit none
   real(kind=8), dimension(9, N, M) :: f
   integer :: i, j, q, k
   real(kind=8) :: s, rho

   s = 0.0d0
   do q=1,9
      s = s + f(q, i, j) * c(q, k)
   end do

   s = s / rho(i, j, f)

end function

subroutine boundary(f)
   use constants, only : c, w
   implicit none
   real(kind=8), dimension(9, N, M) :: f
   real(kind=8) :: ux, s1, s2, cu, rho_j, inlet_vel
   integer :: j, q

   ! outflow
   do q=4,6
      do j=1,M
         f(q, N, j) = f(q, N-1, j)
      end do
   end do

   ! inflow
   do j=1,M
      ux = inlet_vel(j)
      s2 = f(1, 1, j) + f(2, 1, j) + f(3, 1, j)
      s1 = f(4, 1, j) + f(5, 1, j) + f(6, 1, j)

      rho_j = 1.0d0 / (1 - ux) * (s2 + 2*s1)

      do q=7,9
         cu = c(q,1) * ux
         f(q, 1, j) = w(q) * rho_j * (1 + 3*cu + 0.5*9*cu**2 - 0.5*3*ux**2)
      end do
   end do

end subroutine

subroutine collstream(fnew, fold, obstacle, omega)
   use constants, only : c, w, noslip
   implicit none
   real(kind=8), dimension(9, N, M) :: fnew, fold
   integer, dimension(N, M) :: obstacle
   real(kind=8) :: omega, ux, uy, rho_ij, u2, cu
   integer :: i, j, q, pmod

   !$omp parallel do collapse(2) private(i,j,q,rho_ij,ux,uy,u2,cu)
   do j=1,M
      do i=1,N
         rho_ij = 0.0d0
         ux = 0.0d0
         uy = 0.0d0
         do q=1,9
            rho_ij = rho_ij + fold(q,i,j)
            ux = ux + fold(q,i,j) * c(q,1)
            uy = uy + fold(q,i,j) * c(q,2)
         end do
         ux = ux / rho_ij
         uy = uy / rho_ij
         u2 = ux**2 + uy**2
         do q=1,9
            cu = ux*c(q,1) + uy*c(q,2)
            if(obstacle(i,j) == 0) then
               fnew(q, pmod(i + c(q,1) - 1, N) + 1, pmod(j + c(q,2) - 1, M) + 1) =&
                  (1-omega) * fold(q,i,j)&
                  + omega * w(q) * rho_ij * (1 + 3*cu + 0.5*9*cu**2 - 0.5*3*u2)
            else
               fnew(q, pmod(i + c(q,1) - 1, N) + 1, pmod(j + c(q,2) - 1, M) + 1) = fold(noslip(q), i, j)
            end if
         end do
      end do
   end do

end subroutine

subroutine update(fnew, fold, obstacle, omega)
   implicit none
   real(kind=8), dimension(9, N, M) :: fnew, fold
   integer, dimension(N, M) :: obstacle
   real(kind=8) :: omega

   call collstream(fnew, fold, obstacle, omega)
   call boundary(fnew)

end subroutine

subroutine print_u(f, t)
   implicit none
   real(kind=8), dimension(9, N, M) :: f
   integer :: t, i, j
   real(kind=8) :: u

   write (*, '(i0,1x)', advance='no') t
   do i=1,N
      do j=1,M
         write (*, '(f0.10,1x)', advance='no') sqrt(u(1,i,j, f)**2 + u(2,i,j, f)**2)
      end do
   end do
   print *

end subroutine

program lb
   use constants, only : cx, cy, r, c, w
   implicit none
   real(kind=8), parameter :: nu = ULB * r / RE
   real(kind=8), parameter :: omega = 1.0d0 / (3.0d0 * nu + 0.5d0)

   real(kind=8), allocatable :: fnew(:,:,:), fold(:,:,:)
   integer, allocatable :: obstacle(:,:)

   integer :: i, j, q, time
   real(kind=8) :: cu, ux, inlet_vel

   allocate(fnew(9,N,M))
   allocate(fold(9,N,M))
   allocate(obstacle(N,M))

   do j=1,M
      do i=1,N
         if((i-cx)**2 + (j-cy)**2 < r**2) then
            obstacle(i,j) = 1
         else
            obstacle(i,j) = 0
         end if
      end do
   end do

   do j=1,M
      do i=1,N
         ux = inlet_vel(j)
         do q=1,9
            cu = ux * c(q,1)
            fold(q, i, j) = w(q) * (1 + 3*cu + 0.5*9*cu**2 - 0.5*3*ux**2)
         end do
      end do
   end do

   do time=0,T-1
      if(mod(time, 2) == 1) then
         call update(fold, fnew, obstacle, omega)
      else
         if(mod(time, 100) == 0) then
            call print_u(fold, time)
         end if
         call update(fnew, fold, obstacle, omega)
      end if
   end do

end program
